### PR TITLE
Add react-messageformat slide

### PR DESCRIPTION
**Pull request by rickmak**

### CODE REVIEW

File: react-messageformat.slide [hunk 1]
```diff
@@ -0,0 +1,165 @@
+Why Oursky uses its own translation module in react
```
Comment (louischan-oursky) [inline react-messageformat.slide]: I prefer the term "localization" over "translation".

File: react-messageformat.slide [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+* Existing solutions for React translation
+
+- react-intl <- Oursky used this before
+- react-i18next
+- Others?
+
+Let's do a quick poll
+
+* We encountered following problem with react-intl
```
Comment (louischan-oursky) [inline react-messageformat.slide]: "Problems we encountered with react-intl"

File: react-messageformat.slide [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+1. Can't embed react-component in translation string
```
Comment (louischan-oursky) [inline react-messageformat.slide:20]: "Can't embed React Element in translation string"

File: react-messageformat.slide [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+2. ICU(International Components for Unicode) parser differ from the original implementation (icu4c)
+.link https://github.com/yahoo/intl-messageformat
```
Comment (louischan-oursky) [inline react-messageformat.slide:22]: Should we demonstrate a concrete example here?

File: react-messageformat.slide [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+3.  Not yet support new(16.3) Context
+.link https://github.com/yahoo/react-intl/pull/1186
+
+* How is our oursky/react-messageformat different from the others
+
+- Support embedding react-component in ICU messageformat
```
Comment (louischan-oursky) [inline react-messageformat.slide:28]: "Support embedding React Element in ICU message format"

File: react-messageformat.slide [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+- Use a more standard ICU parser
+- Must use 'new' Context API
```
Comment (louischan-oursky) [inline react-messageformat.slide]: "Require the new Context API"

File: react-messageformat.slide [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+- We use github.com/messageformat/parser
+.link https://github.com/oursky/react-messageformat#why
+- We add react directive, add it to parser to make it a first class citizen
+.link https://github.com/louischan-oursky/parser/commit/f9f8d499e2dd696ce9b890dc073950a48f749a09
+
+* React directive of ICU we added
+
+  Please {Link, react, href{http://oursky.com} children{{Text, react, children{Click me}}}}
+
+- `Link` is the name of directive
```
Comment (louischan-oursky) [inline react-messageformat.slide]: "Link is the name of the React Component"

File: react-messageformat.slide [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+- `react` to declare it is react element
```
Comment (louischan-oursky) [inline react-messageformat.slide]: "react is the newly added directive to support React Element"

File: react-messageformat.slide [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+- Others are passed as props, `key{value}`
```
Comment (louischan-oursky) [inline react-messageformat.slide:70]: We can add one more point here to emphasize that the syntax is essentially the same as `select`

File: react-messageformat.slide [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+* Native to new Context API
+
+Just use the following and the FormattedMessage will be using new Context API
+to detect locale change.
+
+  import { Provider, FormattedMessage } from "@oursky/react-messageformat";
+
+* The new context API will save us from
+
+- PureComponent will not refresh on local change. The old Context will not propagate to it.
```
Comment (cychiuae) [inline react-messageformat.slide]: PureComponent will not refresh on **locale** change

File: react-messageformat.slide (hunk 2 above)
Comment (cychiuae) [inline react-messageformat.slide]: problem -> problems

File: react-messageformat.slide [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+ />
+
+* What is ICU adoption?
+
+- i18next also support ICU since 7.6.0, Apr 2018
+- It depends to i18next-icu, which depends on yahoo/intl-messageformat
+- Android provides `android.icu` package since 7.0 (API level 24)
+- Steep learning curve in ICU compared with JSON format
+.link https://github.com/i18next/react-i18next/issues/303#issuecomment-332419079
+- In most case, you just need simple interpolations
```
Comment (cychiuae) [inline react-messageformat.slide]: In most **cases**

File: react-messageformat.slide [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+  en.json
+  {"join": "Click {Bold, react, children{here}} to Join!"}
+  cn.json
+  {"join": "{Bold, react, children{按此}}加入！"}
+
+  import Bold from "./Bold";
+  <FormattedMessage id="join" components={{ Bold }} />
+
+* Native to new Context API
```
Comment (louischan-oursky) [inline react-messageformat.slide]: I would say "Require the new Context API"

File: react-messageformat.slide [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+Just use the following and the FormattedMessage will be using new Context API
+to detect locale change.
+
+  import { Provider, FormattedMessage } from "@oursky/react-messageformat";
```
Comment (louischan-oursky) [inline react-messageformat.slide]: The example here should look like

```typescript
import { Provider } from "@oursky/react-messageformat";
const messages = {
  "Hello.World": "Hello World!",
};

<Provider locale="en" messages={messages}>
  <MyApp />
</Provider>
```

**Review by cychiuae — CHANGES_REQUESTED** [review]

File: react-messageformat.slide (hunk 10 above)
Comment (louischan-oursky) [inline react-messageformat.slide]: "No longer suffer from the broken old Context" Include the link to https://reactjs.org/docs/legacy-context.html#updating-context which explains how it is broken.

File: react-messageformat.slide [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+
+* The new context API will save us from
+
+- PureComponent will not refresh on local change. The old Context will not propagate to it.
+- Work around one: refresh whole app - not ideal in UX.
+.link https://github.com/yahoo/react-intl/issues/243
+- Use Redux to work around the context
```
Comment (louischan-oursky) [inline react-messageformat.slide:149]: Are we going to show how we worked around it?

File: react-messageformat.slide [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+- We will not do all of
+.link https://github.com/yahoo/react-intl/wiki#the-react-intl-module
+- For example, FormattedPlural, FormattedHTMLMessage
+- We rely on the standard Intl too
+.link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl
+
+* Future development
+
+- Support react hook API
+- Support number translate, using standard Intl
```
Comment (louischan-oursky) [inline react-messageformat.slide]: "Support Intl.NumberFormat" Include a link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat

File: react-messageformat.slide [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+- It is not aimed to be a drop-in replacement for yahoo/react-intl
+- We will not do all of
+.link https://github.com/yahoo/react-intl/wiki#the-react-intl-module
+- For example, FormattedPlural, FormattedHTMLMessage
+- We rely on the standard Intl too
+.link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl
+
+* Future development
+
+- Support react hook API
```
Comment (louischan-oursky) [inline react-messageformat.slide:164]: "Support React Hooks API"

File: react-messageformat.slide (hunk 5 above)
Comment (rickmak) [reply react-messageformat.slide:28]: Why React Element over component?

File: react-messageformat.slide (hunk 14 above)
Comment (rickmak) [reply react-messageformat.slide:149]: No. In normal flow.

**Pull request merged**